### PR TITLE
Allow time-based data-splitting with multiple datasources as 'split.basis'

### PR DESCRIPTION
**Pull request by MaLoefUDS**
### Prerequisites

- [x] I adhere to the coding conventions (described [here](../CONTRIBUTING.md)) in my code.
- [x] I have updated the copyright headers of the files I have modified.
- [x] I have written appropriate commit messages, i.e., I have recorded the goal, the need, the needed changes, and the location of my code modifications for each commit. This includes also, e.g., referencing to relevant issues.
- [x] I have put signed-off tags in *all* commits.
- [x] I have updated the changelog file [NEWS.md](../NEWS.md) appropriately.
- [x] I have checked whether I need to adjust the showcase file `showcase.R` with respect to my changes.
- [x] The pull request is opened against the branch `dev`.

### Scope

This PR aims to resolve issue #254

### Description

Allow time-based data-splitting with multiple datasources as `split.basis`

#### Added
* Add test for new functionality

#### Changed
* Adjust handling of `split.basis` in `split.data.by.time.or.bins` and `split.data.time.based` to allow for splitting by multiple datasources

### CODE REVIEW

File: tests/test-split-data-time-based.R [hunk 1]
```diff
@@ -770,6 +770,53 @@ patrick::with_parameters_test_that("Split a data object time-based (split.basis
...
+    expect_equal(min(issue.data$date), as.POSIXct("2016-07-12 15:59:25"))
+    expect_equal(max(issue.data$date), as.POSIXct("2016-07-12 16:06:01"))
+
+    # split by 'mails' and 'issues' without sliding window
+    results = split.data.time.based(project.data, time.period = "1 min",
+                                    split.basis = c("mails", "issues"), sliding.window = test.sliding.window)
+
+    # bins should be union of both sources
+    expect_equal(min(attr(results, "bins")), min(c(issue.data$date, mail.data$date)))
+    expect_equal(max(attr(results, "bins")), max(c(issue.data$date, mail.data$date)) + 1)
```
Comment (bockthom) [inline tests/test-split-data-time-based.R:812]: Hm, I am not sure if it is enough to test for the minimum and maximum? Can it happen that something in between is wrong (i.e., the length of the first range) or is this already ensured by any other test? If so, then please add a comment why it is not necessary to test for the actual bins but just for minimum and maximum.

In addition, as also `split.data.by.bins` is affected by the changes of this PR, could you please also add a suitable test for that? Thanks!

**Review by bockthom — CHANGES_REQUESTED** [review]
Thanks for this PR @MaLoefUDS.

Before I start with my actual review, just a general comment: Could you please reference the issue number somewhere in the PR? Either in the description of the PR or inside the commit messages of suitable commits, such that we have a connection between PR and issue? Thanks!

In general, this PR looks good. I just have a few comments.

File: tests/test-split-data-time-based.R (hunk 1 above)
Comment (MaLoefUDS) [reply tests/test-split-data-time-based.R:812]: Yes, these tests are not as broad as they could be. I did manually check the correctness of the bins, but I have not yet found a good way to unify the test with and without `sliding.windows`, since then we obtain additional (intermediary) bins. Do you have a good idea how to do that? I could do two tests, or just define two bins and decide based on `test.sliding.windows` which one to use?

File: util-split.R [hunk 2]
```diff
@@ -99,6 +108,9 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
 split.data.by.bins = function(project.data, activity.amount, bins, split.basis = c("commits", "mails", "issues"),
                                      sliding.window) {
 
+    ## get basis for splitting process
+    split.basis = match.arg(split.basis)
```
[the remark this answers is not in the document]
Comment (MaLoefUDS) [reply util-split.R:112]: Well that is exactly what I intended. This PR and the issue that it stems from only describe splitting with multiple `split.basis` when splitting time based. Then, **when we create new bins**, we account for multiple datasources. 
When we split by **given bins**, I assumed we do not want to be able to provide bins as input that stem from multiple datasources. Therefore, I repositioned the check that was previously in place to check for that `split.basis` to only includes 1 datasource here in the one wrapper function, while in the other, I included the new check 😄.

**Review by bockthom — APPROVED** [review]
Looks good to me now.

**Comment by MaLoefUDS** [comment]
I will update the `NEWS.md` later today 👍

File: NEWS.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
...
 - Add commit-interaction networks that can be created with `create.author.network` and `create.artifact.network` if the `artifact.relation` and `author.relation` is configured to be `commit.interaction` (PR #252, d82857fbebd1111bb16588a4223bb24a8dcd07de, 329d97ec3de36a9e1bcadc0c7a53c1d92e8b481c) as well as tests for these features (PR #252, 07e7ed744209b0251217fa8f7f35d9b9875face2, 7068cfa10d993dcae3f5e3f76f8cafa99fa8b350)
 - Add helper function for prefixing function names with file names in `util-read.R` (PR #252, f8ea987b138173cf0509c7910e0572d8ee1b3f1f)
+- Add the possibility to split data time-based by multiple data sources and according unit-tests (PR #261, d76aafad74cd9e83ebcc1368f28f77fe63ced7f4, 4d2ebd29f549453f9338116bde59dd327266dac3, b6e99f67640885b3e2f41efbfead79b5d02b5a83, cdbea3c5ce751fca929375215a9be706a6788af4)
```
Comment (bockthom) [inline NEWS.md:12]: Please remove the part "and according unit-tests" - we generally assume that new functionality is properly tested.

File: showcase.R [hunk 4]
```diff
@@ -289,6 +289,14 @@ for (range in names(cf.data)) {
 }
 print(run.lapply(cf.data, "get.class.name"))
 
+cf.data = split.data.time.based(x.data, time.period = "18 month", split.basis = c("commits", "issues"))
```
Comment (bockthom) [inline showcase.R:292]: Could you please add a short comment beforehand? (similar to the comment in line 309).
Maybe it would also make sense to add a comment to the previous splitting functionality above and to the activity-based below in line 318? Thus one line to highlight the differences of the different splitting variants.

**Review by bockthom — CHANGES_REQUESTED** [review]
Looks good to me. I just have two comments regarding comments.

File: NEWS.md (hunk 3 above)
Comment (bockthom) [reply NEWS.md:12]: And one more comment: I'd prefer to merge this PR after the PR that introduces measuring code coverage. This way, we (1) get rid of the failing build pipeline for R 3.6, and (2) we can directly see whether the codecov functionality is properly working on this PR. Therefore, the commit hashes need to be updated after the corresponding rebase.

Comment (MaLoefUDS) [reply NEWS.md:12]: I will therefore wait with fixing those issues in `NEWS.md` until #262 is merged 👍

**Comment by MaLoefUDS** [comment]
🔔 Ping! Do not miss this PR :)

**Comment by bockthom** [comment]
> 🔔 Ping! Do not miss this PR :)

Thanks for the ping. However, this time, the ping was not necessary :smile: Christian and I planned to have a look at this PR today anyway. So, I assume you will either receive comments later today or we will merge it later today.

File: util-split.R [hunk 5]
```diff
@@ -65,6 +66,14 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
                                  number.windows = NULL, split.basis = c("commits", "mails", "issues"),
                                  sliding.window = FALSE, project.conf.new = NULL) {
 
+    # ensure 'split.basis' defaults to 'commits' if not defined
+    # and allow it to contain multiple data sources if explicitly wanted
+    if(!hasArg("split.basis")) {
```
Comment (hechtlC) [inline util-split.R:71]: Space after if

**Review by hechtlC — DISMISSED** [review]
Looks good to me thank you @MaLoefUDS !
I have one formatting comment that you have to address before merging but I approve the PR nonetheless. Please fix this and anything @bockthom might still find.

**Comment by bockthom** [comment]
> Please fix this and anything @bockthom might still find.

There is nothing else to find (at least, for me) :smile: 
So, I will merge right away after the missing space is fixed.

**Review by bockthom — APPROVED** [review]
